### PR TITLE
Hotfix/performance housing fix

### LINKED ISSUE #1492: Unable to reassign colonists to a new house (1.10.2)
### Minecolonies version

version 1.10.2-0.8.4113 or curse 0.8.9

### Expected behavior

- When removing a colonist from their current home, they remain in the list of colonists so you can assign them to another house.

### Actual behaviour

- When removing a colonist from their current home, they seemingly disappear from the list of colonists within the house assignment UI, therefore you are unable to add that colonist to any houses, leaving them homeless. It's also been mentioned by others on the server that any colonists you remove from a house actually return to the house they should have been removed from.

### Steps to reproduce the problem

1. Remove a colonist from a house they are currently assigned to.
2. Search for that colonist within the house assignment UI.
3. ...
4. Profit

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #1492 

# Changes proposed in this pull request:
- Fixes a bug where citizens couldn't be unassigned of their houses
- Fixes an extremely important performance bug where citizen views were updated all the time!
-

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/buildings/BuildingHome.java [hunk 1]
```diff
@@ -164,6 +164,7 @@ public void removeCitizen(@NotNull final CitizenData citizen)
         {
             citizen.setHomeBuilding(null);
             residents.remove(citizen);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/BuildingHome.java:166]: MINOR This call to "remove()" may be a performance hot spot if the collection is large. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2250)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 59 issues
* ![MAJOR][MAJOR] 22 major
* ![MINOR][MINOR] 35 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [CitizenData.java#L26](https://github.com/Minecolonies/minecolonies/blob/8d452ffe34058274311992343fdabf5d2c8a3117/src/main/java/com/minecolonies/coremod/colony/CitizenData.java#L26): class "CitizenData" has 43 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [CitizenData.java#L452](https://github.com/Minecolonies/minecolonies/blob/8d452ffe34058274311992343fdabf5d2c8a3117/src/main/java/com/minecolonies/coremod/colony/CitizenData.java#L452): Use the "equals" method if value comparison was intended. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1698)
1. ![MAJOR][MAJOR] [CitizenData.java#L457](https://github.com/Minecolonies/minecolonies/blob/8d452ffe34058274311992343fdabf5d2c8a3117/src/main/java/com/minecolonies/coremod/colony/CitizenData.java#L457): Use the "equals" method if value comparison was intended. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1698)
1. ![MAJOR][MAJOR] [CitizenData.java#L512](https://github.com/Minecolonies/minecolonies/blob/8d452ffe34058274311992343fdabf5d2c8a3117/src/main/java/com/minecolonies/coremod/colony/CitizenData.java#L512): Use the "equals" method if value comparison was intended. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1698)
1. ![MAJOR][MAJOR] [CitizenData.java#L516](https://github.com/Minecolonies/minecolonies/blob/8d452ffe34058274311992343fdabf5d2c8a3117/src/main/java/com/minecolonies/coremod/colony/CitizenData.java#L516): Use the "equals" method if value comparison was intended. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1698)
1. ![MAJOR][MAJOR] [BuildingHome.java#L330](https://github.com/Minecolonies/minecolonies/blob/8d452ffe34058274311992343fdabf5d2c8a3117/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingHome.java#L330): Reduce this class from 38 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [EntityCitizen.java](https://github.com/Minecolonies/minecolonies/blob/8d452ffe34058274311992343fdabf5d2c8a3117/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java): This file has 1.380 lines, which is greater than 1.000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L71](https://github.com/Minecolonies/minecolonies/blob/8d452ffe34058274311992343fdabf5d2c8a3117/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L71): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 58 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L71](https://github.com/Minecolonies/minecolonies/blob/8d452ffe34058274311992343fdabf5d2c8a3117/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L71): class "EntityCitizen" has 99 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L371](https://github.com/Minecolonies/minecolonies/blob/8d452ffe34058274311992343fdabf5d2c8a3117/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L371): Assign this magic number 3.0F to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/coremod/colony/CitizenData.java [hunk 2]
```diff
@@ -481,9 +481,10 @@ public void setHomeBuilding(@Nullable final AbstractBuilding building)
         if (homeBuilding != null && building != null && !homeBuilding.equals(building))
         {
             homeBuilding.removeCitizen(this);
+            markDirty();
         }
 
-        if (building instanceof BuildingHome || building instanceof BuildingBarracksTower)
+        if (building == null || building instanceof BuildingHome || building instanceof BuildingBarracksTower)
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/colony/CitizenData.java:487]: building == null is irrelevant here and an additional check that is not needed. Null is never an instance of.

**Review by marchermans — CHANGES_REQUESTED** [review]
Remove the null check.

File: src/main/java/com/minecolonies/coremod/colony/CitizenData.java (hunk 2 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/CitizenData.java:487]: Its not, three types of buildings can be assigned to citizens:
BuildingHome, BuildingBarracksTower and null (removing them from a building)

Comment (marchermans) [reply src/main/java/com/minecolonies/coremod/colony/CitizenData.java:487]: Ah sorry misread it.

**Review by marchermans — APPROVED** [review]

**Pull request merged**